doc/source/dev/development_workflow.rst

```diff
@@ -150,6 +153,81 @@ been added to ``upstream`` that affect your work. In this case, follow the
 :ref:`rebasing-on-main` section of this document to apply those changes to
 your branch.
 
+.. _formatting-with-pre-commit:
+
+Automatic code formatting with pre-commit
+-----------------------------------------
+
+NymPy recommends that code follow pep7/pep8 guidelines, but does not currently
+enfoce any formatters. However, the repository does come with a an optional
+pre-commit hook that performs automatic formatting and validation of changes.
+Tasks and checks that it runs include:
+
+* Format changed lines only of Python files to pep8 standard using `Darker
+  <darker>`_, which wraps `black <black>`_ and `isort <isort>`_ (enabled by
+  default). Additionally, check only these changed lines using `flake8
+  <flake8>`_.
+
+* Format changed lines only of C and C++ source files to pep7 standard using a
+  python wrapper for `ClangFormat versioning <clang_format>`_ (enabled by
+  default).
+
+* Format and lint YAML shell files, and perform linting on Fortran, rst, Docker,
+  and markdown files (some enabled and some disabled by default)
+
+To use the hook, first make sure that pre-commit is installed:
+
+:: 
+
+   pip install pre-commit
+
+And then install the hook to your git repository:
+
+:: 
+
+   pre-commit install
+
+This will add a script to your ``.git/hooks`` folder. This script will run
+pre-commit to check your changed files whenever you run ``git commit``. Note
+that if there are any failures, the commit will not go through and you will have
+to re-run the commit (and ``git add`` first if any files were automatically
+changed). It is generally bad practice to commit files that do not pass tests
+but if it is necessary (e.g., for a local commit that will be rebased later),
+you can run ``git commit --no-verify`` to skip pre-commit checks.
+
+Note that the first time pre-commit runs, it will take a while to download the
+necessary files. Subsequent times will be very fast.
```

Comment:
Where do things get downloaded from?

Reply:
Pre-commit downloads from the tagged repo to get the pre commit hook, any dependencies are from PyPi

Reply:
Thanks, that makes sense. Can you check / make sure this doesn't use build isolation or anything like that and uses normal `pip install` commands, so that it works well with a conda environment (like the `numpy-dev` one that's our default, see `environment.yml` in the root of the repo)?

Reply:
Alrightie, I don't really know anaconda so I don't have the best answer but the only dependency is python+pip (verified in a python container that it all works), and pre-commit installed via pip (on the system or in the environment) or via e.g. apt. It only needs to be accessible once on running `pre-commit install` - that adds a script to the git hooks directory that points absolutely to the pre-commit CLI wherever it is.

I don't know enough about build environments to understand what the concern there might be. Would it make sense to add pre-commit to the environment.yml?

Reply:
I had a look, it should work. Also, `pre-commit` does have conda support: https://pre-commit.com/#conda.

Reply:
> I don't know enough about build environments to understand what the concern there might be. Would it make sense to add pre-commit to the environment.yml?

Yes, I think it would. It can be added to the other development dependencies there.